src/vm.rs

```diff
@@ -642,40 +642,51 @@ impl<T: GuestStagePageTable> Vm<T, VmStateFinalized> {
             counter_mask: u64,
             start_flags: PmuCounterStartFlags,
             initial_value: u64,
-            _active_vcpu: &mut ActiveVmCpu<T>,
+            active_vcpu: &mut ActiveVmCpu<T>,
         ) -> EcallResult<u64> {
             let pmu_info = PmuInfo::get()?;
-            let (platform_counter_index, platform_counter_mask) = pmu_info
-                .get_platform_counter_index_and_mask(
-                    LogicalCounterIndex(counter_index),
-                    LogicalCounterMask(counter_mask),
-                )?;
-            pmu::start_counters(
+            let counter_index = LogicalCounterIndex(counter_index);
+            let counter_mask = LogicalCounterMask(counter_mask);
+            let (platform_counter_index, platform_counter_mask) =
+                pmu_info.get_platform_counter_index_and_mask(counter_index, counter_mask)?;
+            active_vcpu.is_counter_range_startable(counter_index, counter_mask)?;
+            let result = pmu::start_counters(
                 platform_counter_index.0,
                 platform_counter_mask.0,
                 start_flags,
                 initial_value,
-            )?;
+            );
+            // Special case "already started" to handle counters that are autostarted following configuration
+            if let Err(e) = result && !matches!(e, SbiError::AlreadyStarted) {
+                return Err(EcallError::from(e));
+            }
```

Comment:
AlreadyStarted error condition should also be propagated to the VM because vm either called start_counter twice or called start_counter even after passing auto start during config.

In fact, you should call update_started_counters if auto start is set during config and it did not return any error. That way, you can just return the error from by checking the mask.

May be I don't get the rust syntax. How are you returning error if the counter is not startable ?